### PR TITLE
Better Publishing and Better Logging

### PR DESCRIPTION
**Pull request by hosamaly**
There is a lot of refactoring in this PR. Highlights include:
* Logging
* Introduction of the `Publisher` trait, making it easy to redirect output elsewhere (e.g. to SFTP).

### CODE REVIEW

File: src/main/scala/com/gu/io/Publisher.scala [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  /** Called before content is persisted */
+  protected def saving(key: String): Unit = { logger.trace(s"Saving $key...") }
+
+  /** Called after content is persisted */
+  protected def saved(fileName: String, link: Link): Unit = {
+    logger.trace(s"Saved $fileName as $link")
+    savedLinks.put(link, null)
+  }
+
+  protected val savedLinks: scala.collection.concurrent.Map[Link, Null] = TrieMap()
```
Comment (tomrf1) [inline src/main/scala/com/gu/io/Publisher.scala]: why is this a map?

File: src/main/scala/com/gu/kindlegen/KindleGenerator.scala [hunk 2]
```diff
@@ -1,116 +1,104 @@
...
+      .flatMap { articles =>
+        Future.sequence(articles.zipWithIndex.map((downloadMainImage _).tupled))
+      }
 
-    (articlesOnDisk ++ articlesOnDisk.flatMap(_.mainImage) ++ sections :+ rootManifest)
-      .map(_.link).collect {
-      case link: PathLink => link.toPath
-    }
+    Await.ready(eventualArticlesWithImages, querySettings.downloadTimeout)  // force the timeout
+      .flatMap { articlesWithImages =>
```
Comment (tomrf1) [inline src/main/scala/com/gu/kindlegen/KindleGenerator.scala]: could a for/yield be neater here?

File: src/main/scala/com/gu/io/Publisher.scala (hunk 1 above)
Comment (hosamaly) [reply src/main/scala/com/gu/io/Publisher.scala]: It was the only concurrent collection type in Scala that I could find!

Comment (tomrf1) [reply src/main/scala/com/gu/io/Publisher.scala]: ugh, that's unfortunate. Worth adding a comment here, to save the next person from wondering the same thing

Comment (hosamaly) [reply src/main/scala/com/gu/io/Publisher.scala]: Done.

**Pull request merged**